### PR TITLE
Fix #570 issue with observed directory deleted

### LINKED ISSUE #570: Python process can not delete folder that is being observed
If you try to delete a folder that is being observed by a python process using watchdog and then try to delete it using another python process the folder becomes unavailable until the watchdog process stops.

Steps to reproduce:

1. Create a folder to observe: **C:\\test_folder**
2. Create a python script using watchdog to observe this folder:
```python
import sys
import time
import logging
from watchdog.observers import Observer
from watchdog.events import LoggingEventHandler

if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO,
                        format='%(asctime)s - %(message)s',
                        datefmt='%Y-%m-%d %H:%M:%S')
    path = 'C:\\test_folder'
    event_handler = LoggingEventHandler()
    observer = Observer()
    observer.schedule(event_handler, path, recursive=True)
    observer.start()
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        observer.stop()
    observer.join()
```

3. Open a python shell and try to delete this folder:
```python
import os
for root, dirs, files in os.walk('C:\\test_folder'):
    for f in files:
    os.remove(os.path.join(root, f))
    os.rmdir('C:\\test_folder')
```

4. Check that the **C:\\test_folder** still appear in file explorer and you are not able to open
5. Stop watchdog process
6. Folder has been deleted

### PR DESCRIPTION
**Pull request by rrzaripov**
This PR provides similar behavior for Windows and Linux in case when observed directory is removing. Windows system call GetFinalPathNameByHandleW used for detection situation, when observed directory deleted. When this situation detected, synthetic event FILE_ACTION_DELETED_SELF generated. WindowsApiEmitter getting this event and stopping thread. Handle to observed directory closes on thread stopping.
TODO: Tests in progress.

### CODE REVIEW

**Comment by BoboTiG** [comment]
Pr good to review or you are still working on it?

**Comment by rrzaripov** [comment]
@BoboTiG this PR ready for review, I still think that this solution can be improved.

File: src/watchdog/observers/winapi.py [hunk 1]
```diff
@@ -301,6 +334,10 @@ def read_directory_changes(handle, recursive):
     except WindowsError as e:
         if e.winerror == ERROR_OPERATION_ABORTED:
             return [], 0
+
+        if _is_observed_path_deleted(handle, path):
```
Comment (BoboTiG) [inline src/watchdog/observers/winapi.py]:
```suggestion
        # Handle the case when the root path is deleted
        if _is_observed_path_deleted(handle, path):
```

**Comment by BoboTiG** [comment]
I like the idea. It is OK to call new functions as they will be used only on specific error. Do you have something else in mind?

**Comment by rrzaripov** [comment]
@BoboTiG only one thing confuses me. When the observed directory was deleted watchdog not sending any event, so user never know that about this. Inotify generate DELETE_SELF event, but it not propagate by watchdog to user. Possible this is good idea for separate PR.

**Comment by BoboTiG** [comment]
Do you mind fixing flake8 errors?

As for the second PR, I think it is better, yes. Watchdog should generate this event for all observers.

**Comment by rrzaripov** [comment]
@BoboTiG fixed, flake8 is really useful! :+1:

**Comment by BoboTiG** [comment]
Thanks :+1:

**Pull request merged**